dpki/src/keystore.rs

```diff
@@ -0,0 +1,330 @@
+    }
+
+    /// adds a random root seed into the keystore
+    #[allow(dead_code)]
+    pub fn add_random_seed(&mut self, id_str: &str, size: usize) -> HcResult<()> {
+        let id = id_str.to_string();
+        if self.keys.contains_key(&id) {
+            return Err(HolochainError::ErrorGeneric(
+                "identifier already exists".to_string(),
+            ));
+        }
+        let seed_buf = generate_random_seed_buf(size);
+        let secret = Arc::new(Mutex::new(Secret::RootSeed(RootSeed::new(seed_buf))));
+        let _ = self.keys.insert(id, secret);
+        Ok(())
+    }
+
+    fn check_dst_identifier(&self, dst_id_str: &str) -> HcResult<String> {
+        let dst_id = dst_id_str.to_string();
+        if self.keys.contains_key(&dst_id) {
+            return Err(HolochainError::ErrorGeneric(
+                "identifier already exists".to_string(),
+            ));
+        }
+        Ok(dst_id)
+    }
+
+    fn check_src_identifier(&self, src_id_str: &str) -> HcResult<Arc<Mutex<Secret>>> {
+        let src_id = src_id_str.to_string();
+        if !self.keys.contains_key(&src_id) {
+            return Err(HolochainError::ErrorGeneric(
+                "unknown source identifier".to_string(),
+            ));
+        }
+        Ok(self.keys.get(&src_id).unwrap().clone()) // unwrap ok because we checked if src exists
+    }
+
+    fn check_identifiers(
+        &self,
+        src_id_str: &str,
+        dst_id_str: &str,
+    ) -> HcResult<(Arc<Mutex<Secret>>, String)> {
+        let dst_id = self.check_dst_identifier(dst_id_str)?;
+        let src_secret = self.check_src_identifier(src_id_str)?;
+        Ok((src_secret, dst_id))
+    }
+
+    /// adds a derived seed into the keystore
+    #[allow(dead_code)]
+    pub fn add_seed_from_seed(
+        &mut self,
+        src_id_str: &str,
+        dst_id_str: &str,
+        context: &SeedContext,
+        index: u64,
+    ) -> HcResult<()> {
+        let (src_secret, dst_id) = self.check_identifiers(src_id_str, dst_id_str)?;
+        let secret = {
+            let mut src_secret = src_secret.lock().unwrap();
+            match *src_secret {
+                Secret::RootSeed(ref mut src) => {
+                    let seed = src.generate_indexed_seed(context, index)?;
+                    Arc::new(Mutex::new(Secret::IndexedSeed(seed)))
+                }
+                _ => {
+                    return Err(HolochainError::ErrorGeneric(
+                        "source secret is not a root seed".to_string(),
+                    ));
+                }
+            }
+        };
+        let _ = self.keys.insert(dst_id, secret);
+
+        Ok(())
+    }
+
+    /// adds a keypair into the keystore based on a seed already in the keystore
+    /// returns the public key
+    #[allow(dead_code)]
+    pub fn add_key_from_seed(
+        &mut self,
+        src_id_str: &str,
+        dst_id_str: &str,
+        context: &SeedContext,
+        index: u64,
```

Comment:
I feel like we shouldn't dual purpose this. If we provide access to discreet functions, then the app dev can make use of them how they want:
- derivation (seed + context + index = new_seed)
- key generation (seed = keypair)

then, the base32 key id (which for an application key is the agentId) could be the key in the keystore that stores the private side)

Reply:
I'm sorry I really don't understand this comment.  Where does this function dual purpose anything?  Is it because I was allowing creating a key from the RootSeed?  That was my attempt at allowing people access to creating keys directly from random_seeds and not having to do derivation if they didn't want to.  But I could change that.